### ISSUE TITLE
Consistently add or don't add `ModelValidationNode` in `ModelBindingResult`s

### ISSUE THREAD

**Issue by dougbu** [aspnet/Mvc#2633]
We have a few inconsistencies in `ModelValidationNode` creation. For example `FormFileModelBinder`  can create `ModelBindingResults` with `ValidationNode != null && !IsModelSet`; most other model binders leave `ValidationNode == null` when `!IsModelSet`. A `ByteArrayModelBinder` counter-example came up in PR #2626, specifically this [comment](https://github.com/aspnet/Mvc/pull/2626#discussion_r31373838) from @harshgMSFT.

Need to define a few guidelines and update the in-the-box binders to follow those guidelines. Guidelines may of course be different for different categories (value provider-using vs. type-matching ...).

Relates to other "do the right thing in MB" issues such as #2445.

**Comment by Eilon** [comment]
If IsModelSet is false, there won't be a ModelValidationNode, but there will be a ModelState error stating that the field is required.

**Comment by dougbu** [comment]
Looks like `HttpRequestMessageModelBinder` needs to change and a couple of other classes could change.

Separately `ModelBindingHelper.TryUpdateModelAsync()` completely ignores the `ModelValidationNode` tree the model binding system returns. This method does not even check `ModelBindingResult.IsModelSet` before using `...Model`. Makes careful `ModelValidationNode` creation irrelevant in scenarios involving `Controller.TryUpdateModelAsync()`.

TL;DR
Full list of oddities:
- `CancellationTokenModelBinder`, `FormCollectionModelBinder`, and `FormFileModelBinder` include a `ModelValidationNode` in their successful `ModelBindingResult` instances. These `ModelValidationNode` instances could have `SuppressValidation = true` for consistency with default `MvcOptions.ValidationExcludeFilters` entries. @rynowak thoughts?
- `CollectionModelBinder` _looks_ like it can create a `ModelBindingResult` with `IsModelSet == true` but `ValidationNode == null`. Fortunately the `null` propagating expression that misleads the reader never evaluates to `null`.
- `HttpRequestMessageModelBinder` does not provide a `ModelValidationNode` when `IsModelSet` is `true`. Should probably suppress validation in that case instead of validating the `HttpRequestMessage` and its properties.
  - Oddly `WebApiCompatShimOptionsSetup.Configure()` does not include `typeof(HttpRequestMessage)` in `MvcOptions.ValidationExcludeFilters`. That may be good to add as well.
- `TypeConverterModelBinder` creates a `ModelValidationNode` even when `IsModelSet` is `false`. But at least `DefaultControllerActionArgumentBinder.BindModelAsync()` ignores the `ModelBindingResult.ValidationNode` property when `...IsModelSet` is `false`. So extra `ModelValidationNode` here is just extra memory.

**Comment by dougbu** [comment]
@Eilon @rynowak investigation complete. Do we need these quick fixes in Beta 7?

**Comment by Eilon** [comment]
@dougbu please review https://en.wikipedia.org/wiki/TL;DR :smile:

**Comment by Eilon** [comment]
Let's the three of us chat tomorrow morning about whether to take this for beta7.

**Comment by Eilon** [comment]
1. For model binders that are used for non-validatable types we should exclude them from validation by creating validation nodes.
2. TryUpdateModel should have the same behavior (within its constraints) as regular model binding. Please move this to a new bug.

**Comment by dougbu** [comment]
#2941 covers (2).

**Comment by dougbu** [comment]
83a559c
